### PR TITLE
Room and building acoustics: ISO 18233, ISO 3382-1/2/3, ISO 16283-1 + ISO 717-1

### PR DESCRIPTION
**Pull request by jmrplens**
Closes the full room/building measurement chain: modern impulse-response acquisition, room acoustic parameters, open-plan office spatial metrics (multi-microphone), and field airborne sound insulation with single-number ratings.

## New modules

- **`room_ir`** — ISO 18233:2006: `sweep_signal` (exponential sine sweep, exact analytic phase), `inverse_filter` (Farina), `impulse_response` (spectral deconvolution + Farina harmonic-separation option), `mls_signal` (LFSR, orders 2–20, published primitive-polynomial taps), `mls_impulse_response` (circular cross-correlation with synchronous multi-period averaging).
- **`room_acoustics`** — ISO 3382-1:2009 / 3382-2:2008: `decay_curve` (Schroeder backward integration with noise truncation + tail compensation) and `room_parameters` → per-band EDT/T20/T30 (least-squares over the normative evaluation ranges), C50/C80, D50, Ts, with dynamic-range validity flags and decay curvature.
- **`open_plan`** — ISO 3382-3:2012: `open_plan_metrics` → D2,S and Lp,A,S,4m (log-distance regression, 2–16 m window), distraction/privacy distances rD/rP (STI 0.50/0.20).
- **`insulation`** — ISO 16283-1:2014 + ISO 717-1: `airborne_insulation` (D, DnT, R' with energetic position averaging), `weighted_rating` (reference-curve shifting for Rw/DnT,w with C/Ctr adaptation terms, third-octave and octave band sets), `energy_average_level`.

## Verification

- Closed-form conformance throughout: exponential IR → EDT=T20=T30=T (±1 %); C50/C80/D50/Ts against derived closed forms; known IIR recovered within 0.1 dB in band by both sweep and MLS chains; −6 dB/doubling field → D2,S = 6.0 exact; ISO 717-1 **Annex C worked example reproduced exactly** (Rw(C;Ctr) = 30(−2;−3), unfavourable sum 31.8 dB) plus exact-32.0 dB edge cases; ISO 717-1:2013 cross-checked against the 2020 edition (identical tables).
- Round-half-up implemented explicitly (numpy's banker's rounding avoided where the standard prescribes otherwise).
- 488 tests passing (+71 on this branch), mypy --strict, ruff, bandit clean.

## Docs

- New bilingual guide **Room & building acoustics** (EN/ES) with theory, executed snippets and parameter tables; theory and API reference extended (EN/ES); landing pages updated (19 standards).
- 4 new figure sets ×4 variants (EN/ES × light/dark), all visually audited: Schroeder decay with T20/T30 fits, ISO 717-1 rating with shaded unfavourable deviations, ISO 16283-1 setup diagram, ISO 18233 measurement-chain diagram. Figure annotations are computed by the library itself.
- Markdown/README parity pass: docs/ twins aligned with the site, GitHub-safe math scrubbed in two older guides, llms.txt coverage extended from 11 to 14 pages.
- `generate_graphs.py`/`generate_diagrams.py` moved to `scripts/` with the rest of the dev tooling.

Note: raw image URLs point at `main` and will 404 until this PR merges (intentional, same as previous feature PRs).

https://claude.ai/code/session_013kkVt3nxi9svp1an28uHxf

<a href="https://cubic.dev/pr/jmrplens/phonometry/pull/81?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added room & building acoustics capabilities: deterministic excitation/impulse-response acquisition, decay/room-parameter analysis, open-plan speech privacy metrics, and airborne insulation with weighted single-number ratings.
  * Expanded documentation with a new “Room & building acoustics” guide (EN/ES), updated API reference entries, and refreshed the highlights/navigation.
  * Added new example diagrams/graphs for acoustics workflows.
* **Bug Fixes**
  * Updated calibration validation to IEC 60942:2017 and improved frequency-dependent sensitivity handling.
* **Chores**
  * Improved documentation generation commands/paths for graph/diagram creation.

### CODE REVIEW

File: src/phonometry/room_ir.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+    inv = sweep[::-1] * env
+    # Normalise so that the compressed pulse (sweep * inverse) has unit
+    # magnitude across the sweep band. Scaling by the in-band gain (rather
+    # than the pulse peak) makes the deconvolution reproduce a system's true
+    # in-band level, matching the spectral-division convention.
+    comp = signal.fftconvolve(sweep, inv)
+    freqs = np.fft.rfftfreq(comp.size, d=1.0 / fs)
+    band = (freqs >= f1) & (freqs <= f2)
+    gain = float(np.median(np.abs(np.fft.rfft(comp)[band])))
+    if gain != 0.0:
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room_ir.py]: Fixed in e4c8c25 — an all-False band now raises a clear `ValueError` (increase 'seconds' or widen the range) instead of propagating NaN; chose the explicit error over a closest-bin fallback because a sweep too short to resolve its own band is a measurement-setup mistake the user should see. Regression test added, including propagation through `impulse_response(method='farina')`.

File: src/phonometry/room_ir.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+        inv = inv / gain
+    return inv
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room_ir.py:204]: Fixed in e4c8c25 exactly as proposed — `ValueError` on empty band; the guard is now `gain > 0.0` (also clears the SonarCloud float-equality bug on the same line). Regression test added.

File: src/phonometry/room_acoustics.py [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    t_smooth = (np.arange(smoothed.size) + 0.5 * window) / fs
+    tiny = np.finfo(np.float64).tiny
+    level = 10.0 * np.log10(np.maximum(smoothed, tiny))
+    noise_db = 10.0 * np.log10(noise_power)
+    mask = (level <= level.max() - 5.0) & (level >= noise_db + _TRUST_MARGIN_DB)
+    if int(mask.sum()) < 2:
+        return no_truncation
+    slope, intercept = np.polyfit(t_smooth[mask], level[mask], 1)
+    if slope >= 0.0:
+        return no_truncation
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room_acoustics.py:170]: Fixed in e4c8c25 — guard changed to `slope >= -1e-7` (a slope of −1e−7 dB/s corresponds to T60 ≈ 6·10⁸ s, physically meaningless), with a comment stating the rationale. Regression test: a no-decay noise 'IR' returns finite results with validity flags False and no RuntimeWarning.

File: src/phonometry/room_ir.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+    out_len = length if length is not None else rec.size
+    n = rec.size + ref.size - 1
+
+    if method == "spectral":
+        spec_x = np.fft.rfft(ref, n=n)
+        spec_y = np.fft.rfft(rec, n=n)
+        power = np.abs(spec_x) ** 2
+        reg = regularization * float(power.max()) if power.size else 0.0
+        h_spec = spec_y * np.conj(spec_x) / (power + reg)
+        full = np.fft.irfft(h_spec, n=n)
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room_ir.py:268]: Fixed in e4c8c25 — `impulse_response` now validates both inputs are 1-D and non-empty with clear `ValueError`s, matching the package's validation style. Tests for 2-D and empty on both arguments.

File: src/phonometry/room_ir.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        length ``2**N - 1``.
+    :return: The recovered impulse response.
+    """
+    rec = _typesignal(recorded)
+    seq = _typesignal(mls)
+    period = seq.size
+    if rec.size == 0 or rec.size % period != 0:
+        raise ValueError(
+            "recorded length must be a positive multiple of the MLS length"
+        )
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room_ir.py:354]: Fixed in e4c8c25 — same validation added to `mls_impulse_response` for both `recorded` and `mls`, with tests.

File: src/phonometry/room_ir.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+) -> np.ndarray:
+    """
+    Build the Farina inverse filter for an exponential sine sweep.
+
+    The inverse filter is the time-reversed sweep multiplied by an amplitude
+    envelope that rises by 6 dB/octave (``prop. to the instantaneous
+    frequency``), which whitens the ESS's pink (-3 dB/octave) spectrum so
+    that convolving the sweep with its inverse yields an impulse
+    (ISO 18233:2006, B.5, Figure B.2; Farina 2000, Bibliography [14]). The
+    filter is scaled so that ``sweep * inverse`` has unit peak.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/room_ir.py]: Fixed in e4c8c25 — the docstring now describes the actual convention: normalization by the median in-band magnitude of the compressed pulse so deconvolution reproduces a system's true in-band level.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -53,14 +53,14 @@ pytest --cov=src/phonometry --cov-report=term-missing tests/
...
+`scripts/generate_graphs.py` — never hand-made. If your change alters filter responses,
 weighting curves or any other plotted behavior, regenerate the graphs and commit
 the affected images together with the code change:
 ```bash
-make graphs   # or: python generate_graphs.py
+make graphs   # or: python scripts/generate_graphs.py
 ```
 When adding a feature with visual output, add a `generate_*` function to
-`generate_graphs.py` and reference the resulting image from the docs — do not
+`scripts/generate_graphs.py` and reference the resulting image from the docs — do not
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply CONTRIBUTING.md]: Fixed in e4c8c25 — CONTRIBUTING now names both generators and notes `make graphs` runs the two.

File: docs/room-acoustics.md [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+     28.0, 30.5, 31.8, 32.5, 33.4, 33.0, 31.0, 25.5]
+w = weighted_rating(R)
+print(w.rating, w.c, w.ctr)                          # 30 -2 -3  ->  Rw(C;Ctr) = 30(-2;-3)
+```
+
+To feed `weighted_rating`, the `t2` and level spectra must be the 16
+one-third-octave bands from 100 Hz to 3150 Hz — for example
+`room_parameters(ir, fs, limits=(100, 3150), fraction=3)` — so that every
+band aligns index-by-index with the ISO 717-1 reference curve.
+
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply docs/room-acoustics.md]: Right call, thanks — the sentence conflated obtaining `t2` with feeding `weighted_rating`. Fixed in e4c8c25 (EN docs, EN site, ES site, llms-full): the pipeline now reads l1/l2/t2 on the 16-band 100–3150 Hz grid (t2 e.g. `room_parameters(ir, fs, limits=(100, 3150), fraction=3).t30`) → `airborne_insulation` → its `dnt`/`r_prime` into `weighted_rating`.

File: site/src/content/docs/index.mdx [hunk 9]
```diff
@@ -29,4 +29,7 @@ import { CardGrid, Card } from '@astrojs/starlight/components';
   <Card title="Field measurement" icon="rocket">
     p-p sound intensity with ISO 9614-1 field indicators, calibrated SPL with IEC 60942 stability checks, stateful real-time block processing and vectorized multichannel.
   </Card>
+  <Card title="Room & building acoustics" icon="open-book">
+    Swept-sine and MLS impulse responses (ISO 18233), reverberation and room parameters EDT/T20/T30/C50/C80/Ts (ISO 3382-1/2), open-plan speech metrics (ISO 3382-3) and field sound insulation with Rw/DnT,w and C/Ctr (ISO 16283-1, ISO 717-1).
+  </Card>
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply site/src/content/docs/index.mdx:34]: Fixed in e4c8c25 — the card now says R′w/DnT,w (field ratings), matching the guide's terminology; same fix in the ES landing and README.

File: tests/test_open_plan.py [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+    """<2 positions in 2-16 m -> D2,S / Lp,A,S,4m NaN, rD/rP still work."""
+    r = np.array([0.5, 1.0, 1.5, 18.0])  # all below 2 m or above 16 m: none in range
+    lp = np.array([60.0, 58.0, 56.0, 40.0])
+    sti = 0.6 - 0.02 * r
+    res = open_plan_metrics(r, lp, sti)
+    assert math.isnan(res.d2s)
+    assert math.isnan(res.lp_as_4m)
+    # STI regression uses all positions: 0.5 = 0.6 - 0.02 r -> r = 5
+    assert res.rd == pytest.approx(5.0, abs=1e-9)
+
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply tests/test_open_plan.py:122]: Added in e4c8c25 — `test_exactly_one_position_in_range`: with one position in the 2–16 m window the regression degrades to NaN (the module's documented behavior for <2 usable positions), asserted explicitly.

File: scripts/generate_graphs.py [hunk 11]
```diff
@@ -1577,6 +1597,160 @@ def standing_pressure(pos: float) -> np.ndarray:
...
+    """Schroeder backward integration with T20/T30/EDT regressions (ISO 3382)."""
+    print("Generating schroeder_decay.png...")
+    from phonometry import decay_curve, room_parameters
+    from phonometry.room_acoustics import (
+        _EDT_RANGE,
+        _T20_RANGE,
+        _T30_RANGE,
+        _onset_index,
+    )
+
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply scripts/generate_graphs.py:1610]: Keeping as is, deliberately: the figure generator imports the library's private evaluation-range constants so the published figure can never drift from the code — if the T20/T30 windows ever changed, the figure follows automatically. It is dev tooling inside the same repo, and the graph-guard tests (tests/test_graph_measurements.py) exist precisely to keep the generators honest. A duplicated hardcoded copy of the ranges would be the real maintainability risk.

File: src/phonometry/insulation.py [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+        (Clause 3.13, Formula (2)).
+    :ivar r_prime: Apparent sound reduction index ``R'`` per band, in dB
+        (Clause 3.14, Formula (4)), or ``None`` when the partition area
+        and receiving-room volume were not supplied.
+    """
+
+    d: np.ndarray
+    dnt: np.ndarray
+    r_prime: np.ndarray | None
+
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply src/phonometry/insulation.py:94]: Keeping as is for this PR: all seven result dataclasses in the package (STIResult, IntensityResult, RoomAcousticsResult, OpenPlanResult, both insulation results…) share this exact frozen-with-ndarrays pattern, and comparing/hashing result objects is not a supported use. Changing only this one would break the package-wide symmetry; if eq/hash semantics are ever wanted, that's a one-shot package-wide pass (eq=False or tuple conversion) better done outside a feature PR.

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -53,15 +53,16 @@ pytest --cov=src/phonometry --cov-report=term-missing tests/
...
 the affected images together with the code change:
 ```bash
-make graphs   # or: python generate_graphs.py
+make graphs   # runs both: python scripts/generate_graphs.py && python scripts/generate_diagrams.py
 ```
 When adding a feature with visual output, add a `generate_*` function to
-`generate_graphs.py` and reference the resulting image from the docs — do not
-commit images produced any other way.
+`scripts/generate_graphs.py` (or `scripts/generate_diagrams.py` for a diagram) and
+reference the resulting image from the docs — do not commit images produced any other way.
```
[the remark this answers is not in the document]
Comment (jmrplens) [reply CONTRIBUTING.md:67]: Fixed in c5252e7.
